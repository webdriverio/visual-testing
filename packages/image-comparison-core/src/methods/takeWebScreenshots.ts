import { takeBase64BiDiScreenshot, takeBase64Screenshot } from './screenshots.js'
import { makeCroppedBase64Image } from './images.js'
import { determineScreenRectangles } from './rectangles.js'
import type { WebScreenshotDataOptions, WebScreenshotData } from './screenshots.interfaces.js'
import type { ScreenRectanglesOptions } from './rectangles.interfaces.js'

export async function takeWebScreenshot(
    browserInstance: WebdriverIO.Browser,
    options: WebScreenshotDataOptions,
    shouldUseBidi: boolean = false
): Promise<WebScreenshotData> {
    let base64Image: string

    if (shouldUseBidi) {
        // Take the screenshot with the BiDi method
        base64Image = await takeBase64BiDiScreenshot({ browserInstance })
    } else {
        // Take the screenshot with the regular method
        base64Image = await takeBase64Screenshot(browserInstance)

        // Determine the rectangles
        const screenRectangleOptions: ScreenRectanglesOptions = {
            devicePixelRatio: options.devicePixelRatio || NaN,
            enableLegacyScreenshotMethod: options.enableLegacyScreenshotMethod,
            innerHeight: options.innerHeight || NaN,
            innerWidth: options.innerWidth || NaN,
            isAndroidChromeDriverScreenshot: options.isAndroidChromeDriverScreenshot,
            isAndroidNativeWebScreenshot: options.isAndroidNativeWebScreenshot,
            isEmulated: options.isEmulated || false,
            initialDevicePixelRatio: options.initialDevicePixelRatio || NaN,
            isIOS: options.isIOS,
            isLandscape: options.isLandscape,
        }
        const rectangles = determineScreenRectangles(base64Image, screenRectangleOptions)

        // Make a cropped base64 image
        base64Image = await makeCroppedBase64Image({
            addIOSBezelCorners: options.addIOSBezelCorners,
            base64Image,
            deviceName: options.deviceName,
            devicePixelRatio: options.devicePixelRatio || NaN,
            isIOS: options.isIOS,
            isLandscape: options.isLandscape,
            rectangles,
        })
    }

    return {
        base64Image,
    }
}
