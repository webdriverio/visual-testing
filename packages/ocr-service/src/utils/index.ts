import { join } from 'node:path'
import { mkdirSync } from 'node:fs'
import type { ChainablePromiseElement } from 'webdriverio'
import type { ClickPoint, DetermineClickPointOptions, Rectangles, RectReturn, ScreenshotSize } from '../types.js'

export function getBase64ScreenshotSize(screenshot: string): ScreenshotSize {
    return {
        height: Math.round(Buffer.from(screenshot, 'base64').readUInt32BE(20)),
        width: Math.round(Buffer.from(screenshot, 'base64').readUInt32BE(16)),
    }
}

export function getDprPositions(values: Rectangles, dpr: number): Rectangles {
    Object.keys({ ...values }).map((value: string) => {
    // @ts-ignore
        values[value] /= dpr
    })

    return values
}

/**
 * Determine the click point
 */
export function determineClickPoint(options: DetermineClickPointOptions): ClickPoint {
    const { rectangles: { left, right, top, bottom } } = options
    const x = Math.round( left + (right - left) / 2 )
    const y = Math.round( top + (bottom - top) / 2 )

    return { x, y }
}

/**
 * Adjust the Element Bbox
 */
export function adjustElementBbox(bbox: Rectangles, elementRect: RectReturn): Rectangles {
    return {
        left: Math.round(bbox.left + elementRect.x),
        top: Math.round(bbox.top + elementRect.y),
        right: Math.round(bbox.right + elementRect.x),
        bottom: Math.round(bbox.bottom + elementRect.y),
    }
}

export function createOcrDir(imagesPath: string): string {
    const ocrDir =  join(imagesPath, 'ocr')
    mkdirSync(ocrDir, { recursive: true })

    return ocrDir
}

export function isRectanglesObject(obj: WebdriverIO.Element | ChainablePromiseElement | RectReturn): boolean {
    const properties = ['x', 'y', 'width', 'height']

    return obj !== null &&
           typeof obj === 'object' &&
           properties.every(prop => typeof (obj as any)[prop] === 'number' && !isNaN((obj as any)[prop]) && (obj as any)[prop] >= 0)
}
