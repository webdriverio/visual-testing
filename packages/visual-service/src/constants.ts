export const V6_CLIP_SELECTOR = '#root > :first-child:not(script):not(style)'
export const CLIP_SELECTOR = '#storybook-root > :first-child:not(script):not(style)'
export const NUM_SHARDS = 1
export const PAGE_OPTIONS_MAP: { [key: string]: string } = {
    'saveScreen': 'saveScreenOptions',
    'saveFullPageScreen': 'saveFullPageOptions',
    'saveTabbablePage': 'saveTabbableOptions',
    'checkScreen': 'checkScreenOptions',
    'checkFullPageScreen': 'checkFullPageOptions',
    'checkTabbablePage': 'checkTabbableOptions'
}
