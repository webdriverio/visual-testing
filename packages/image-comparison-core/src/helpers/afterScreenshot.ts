import { join } from 'node:path'
import logger from '@wdio/logger'
import hideScrollBars from '../clientSideScripts/hideScrollbars.js'
import removeElementFromDom from '../clientSideScripts/removeElementFromDom.js'
import { CUSTOM_CSS_ID } from './constants.js'
import { formatFileName, getAndCreatePath } from './utils.js'
import { saveBase64Image } from '../methods/images.js'
import type { AfterScreenshotOptions, ScreenshotOutput } from './afterScreenshot.interfaces.js'
import hideRemoveElements from '../clientSideScripts/hideRemoveElements.js'
import toggleTextTransparency from '../clientSideScripts/toggleTextTransparency.js'

const log = logger('@wdio/visual-service:@wdio/image-comparison-core:afterScreenshot')

/**
 * Methods that need to be executed after a screenshot has been taken
 * to set all back to the original state
 */
export default async function afterScreenshot(browserInstance: WebdriverIO.Browser, options: AfterScreenshotOptions): Promise<ScreenshotOutput> {
    const {
        actualFolder,
        base64Image,
        disableBlinkingCursor,
        disableCSSAnimation,
        enableLayoutTesting,
        fileName: fileNameOptions,
        filePath,
        hideElements,
        hideScrollBars: noScrollBars,
        isLandscape,
        isNativeContext,
        removeElements,
    } = options
    const path = getAndCreatePath(actualFolder, filePath)
    const fileName = formatFileName(fileNameOptions)

    await saveBase64Image(base64Image, join(path, fileName))

    const result = {
        devicePixelRatio: fileNameOptions.devicePixelRatio,
        fileName,
        isLandscape,
        path,
    }

    if (isNativeContext) {
        return result
    }

    if (noScrollBars) {
        await browserInstance.execute(hideScrollBars, !noScrollBars)
    }

    if ((hideElements && hideElements.length > 0) || (removeElements && removeElements.length > 0)) {
        try {
            await browserInstance.execute(hideRemoveElements, { hide: hideElements, remove: removeElements }, false)
        } catch (e) {
            log.warn(
                '\x1b[33m%s\x1b[0m',
                `
#####################################################################################
 WARNING:
 (One of) the elements that needed to be hidden or removed could not be found on the
 page and caused this error
 Error: ${e}
 We made sure the test didn't break.
#####################################################################################
`,
            )
        }
    }

    if (disableCSSAnimation || disableBlinkingCursor || browserInstance.isMobile) {
        await browserInstance.execute(removeElementFromDom, CUSTOM_CSS_ID)
    }

    if (enableLayoutTesting){
        await browserInstance.execute(toggleTextTransparency, !enableLayoutTesting)
    }

    return result
}
