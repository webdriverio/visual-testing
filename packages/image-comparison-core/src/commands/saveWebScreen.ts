import { takeWebScreenshot } from '../methods/takeWebScreenshots.js'
import beforeScreenshot from '../helpers/beforeScreenshot.js'
import afterScreenshot from '../helpers/afterScreenshot.js'
import type { BeforeScreenshotResult } from '../helpers/beforeScreenshot.interfaces.js'
import type { ScreenshotOutput } from '../helpers/afterScreenshot.interfaces.js'
import type { InternalSaveScreenMethodOptions } from './save.interfaces.js'
import type { WebScreenshotDataOptions } from '../methods/screenshots.interfaces.js'
import { canUseBidiScreenshot, getMethodOrWicOption } from '../helpers/utils.js'
import { createBeforeScreenshotOptions, buildAfterScreenshotOptions } from '../helpers/options.js'

/**
 * Saves an image of the viewport of the screen
 */
export default async function saveWebScreen(
    {
        browserInstance,
        instanceData,
        folders,
        tag,
        saveScreenOptions,
        isNativeContext = false,
    }: InternalSaveScreenMethodOptions
): Promise<ScreenshotOutput> {
    // 1. Set some variables
    const { addIOSBezelCorners, formatImageName, savePerInstance } = saveScreenOptions.wic
    const enableLegacyScreenshotMethod = getMethodOrWicOption(saveScreenOptions.method, saveScreenOptions.wic, 'enableLegacyScreenshotMethod')

    // 2.  Prepare the screenshot
    const beforeOptions = createBeforeScreenshotOptions(instanceData, saveScreenOptions.method, saveScreenOptions.wic)
    const enrichedInstanceData: BeforeScreenshotResult = await beforeScreenshot(browserInstance, beforeOptions)
    const {
        deviceName,
        dimensions: {
            window: {
                devicePixelRatio,
                innerHeight,
                innerWidth,
                isEmulated,
                isLandscape,
            },
        },
        initialDevicePixelRatio,
        isAndroid,
        isAndroidChromeDriverScreenshot,
        isAndroidNativeWebScreenshot,
        isIOS,
        isMobile,
    } = enrichedInstanceData

    // 3. Take the screenshot
    const shouldUseBidi = canUseBidiScreenshot(browserInstance) && !isMobile && !enableLegacyScreenshotMethod
    const webScreenshotOptions: WebScreenshotDataOptions = {
        addIOSBezelCorners,
        deviceName,
        devicePixelRatio: devicePixelRatio || 1,
        enableLegacyScreenshotMethod,
        innerHeight,
        innerWidth,
        initialDevicePixelRatio,
        isAndroid,
        isAndroidChromeDriverScreenshot,
        isAndroidNativeWebScreenshot,
        isEmulated,
        isIOS,
        isLandscape,
        isMobile,
    }
    const { base64Image } = await takeWebScreenshot(browserInstance, webScreenshotOptions, shouldUseBidi)

    // 4. Return the data
    const afterOptions = buildAfterScreenshotOptions({
        base64Image,
        folders,
        tag,
        isNativeContext,
        instanceData,
        enrichedInstanceData,
        beforeOptions,
        wicOptions: { formatImageName, savePerInstance }
    })

    return afterScreenshot(browserInstance, afterOptions)
}
