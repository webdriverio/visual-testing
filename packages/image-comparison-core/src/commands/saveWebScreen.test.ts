import { describe, it, expect, vi, afterEach } from 'vitest'
import saveWebScreen from './saveWebScreen.js'
import { takeWebScreenshot } from '../methods/takeWebScreenshots.js'
import afterScreenshot from '../helpers/afterScreenshot.js'
import { canUseBidiScreenshot } from '../helpers/utils.js'
import { createBeforeScreenshotOptions } from '../helpers/options.js'
import type { InternalSaveScreenMethodOptions } from './save.interfaces.js'
import {
    BASE_CHECK_OPTIONS,
    BEFORE_SCREENSHOT_OPTIONS,
    createMethodOptions,
    createTestOptions,
    createBeforeScreenshotMock
} from '../mocks/mocks.js'
import { DEVICE_RECTANGLES } from '../helpers/constants.js'

vi.mock('../methods/takeWebScreenshots.js', () => ({
    takeWebScreenshot: vi.fn().mockResolvedValue({
        base64Image: 'web-screenshot-data'
    })
}))
vi.mock('../helpers/beforeScreenshot.js', () => ({
    default: vi.fn().mockResolvedValue({
        browserName: 'chrome',
        browserVersion: '120.0.0',
        deviceName: 'desktop',
        dimensions: {
            body: {
                scrollHeight: 1000,
                offsetHeight: 1000
            },
            html: {
                clientWidth: 1200,
                scrollWidth: 1200,
                clientHeight: 900,
                scrollHeight: 1000,
                offsetHeight: 1000
            },
            window: {
                devicePixelRatio: 2,
                innerHeight: 900,
                innerWidth: 1200,
                isEmulated: false,
                isLandscape: false,
                outerHeight: 1000,
                outerWidth: 1200,
                screenHeight: 1080,
                screenWidth: 1920,
            },
        },
        isAndroid: false,
        isAndroidChromeDriverScreenshot: false,
        isAndroidNativeWebScreenshot: false,
        isIOS: false,
        isMobile: false,
        isTestInBrowser: true,
        isTestInMobileBrowser: false,
        addressBarShadowPadding: 0,
        toolBarShadowPadding: 0,
        appName: '',
        logName: 'chrome',
        name: 'chrome',
        platformName: 'desktop',
        platformVersion: '120.0.0',
        devicePixelRatio: 2,
        deviceRectangles: {
            bottomBar: { height: 0, width: 0, x: 0, y: 0 },
            homeBar: { height: 0, width: 0, x: 0, y: 0 },
            leftSidePadding: { height: 0, width: 0, x: 0, y: 0 },
            rightSidePadding: { height: 0, width: 0, x: 0, y: 0 },
            screenSize: { height: 0, width: 0 },
            statusBar: { height: 0, width: 0, x: 0, y: 0 },
            statusBarAndAddressBar: { height: 0, width: 0, x: 0, y: 0 },
            viewport: { height: 0, width: 0, x: 0, y: 0 }
        },
        initialDevicePixelRatio: 2,
        nativeWebScreenshot: false
    })
}))
vi.mock('../helpers/afterScreenshot.js', () => ({
    default: vi.fn().mockResolvedValue({
        devicePixelRatio: 2,
        fileName: 'test-screen.png'
    })
}))
vi.mock('../helpers/utils.js', () => ({
    canUseBidiScreenshot: vi.fn().mockReturnValue(false),
    getMethodOrWicOption: vi.fn().mockImplementation((method, wic, option) => method[option] ?? wic[option])
}))
vi.mock('../helpers/options.js', async (importOriginal) => {
    const actual = await importOriginal() as any
    return {
        ...actual,
        createBeforeScreenshotOptions: vi.fn().mockReturnValue({
            instanceData: { test: 'data' },
            addressBarShadowPadding: 6,
            toolBarShadowPadding: 6,
            disableBlinkingCursor: false,
            disableCSSAnimation: false,
            enableLayoutTesting: false,
            hideElements: [],
            noScrollBars: true,
            removeElements: [],
            waitForFontsLoaded: false,
        }),
    }
})

describe('saveWebScreen', () => {
    const takeWebScreenshotSpy = vi.mocked(takeWebScreenshot)
    const afterScreenshotSpy = vi.mocked(afterScreenshot)
    const canUseBidiScreenshotSpy = vi.mocked(canUseBidiScreenshot)
    const createBeforeScreenshotOptionsSpy = vi.mocked(createBeforeScreenshotOptions)

    const baseOptions = {
        browserInstance: { isAndroid: false, isMobile: false } as any,
        folders: BASE_CHECK_OPTIONS.folders,
        instanceData: {
            ...BEFORE_SCREENSHOT_OPTIONS.instanceData,
            deviceRectangles: {
                ...DEVICE_RECTANGLES,
                screenSize: {
                    width: 1920,
                    height: 1080
                }
            }
        },
        isNativeContext: false,
        saveScreenOptions: {
            wic: BASE_CHECK_OPTIONS.wic,
            method: createMethodOptions({
                disableBlinkingCursor: false,
                disableCSSAnimation: false,
                enableLayoutTesting: false,
                enableLegacyScreenshotMethod: false,
                hideScrollBars: true,
                hideElements: [],
                removeElements: [],
                waitForFontsLoaded: true,
            })
        },
        tag: 'test-screen'
    } as InternalSaveScreenMethodOptions

    afterEach(() => {
        vi.clearAllMocks()
    })

    it('should call takeWebScreenshot with correct options when BiDi is available', async () => {
        canUseBidiScreenshotSpy.mockReturnValueOnce(true)
        const result = await saveWebScreen(baseOptions)

        expect(result).toMatchSnapshot()
        expect(createBeforeScreenshotOptionsSpy).toHaveBeenCalledWith(
            baseOptions.instanceData,
            baseOptions.saveScreenOptions.method,
            baseOptions.saveScreenOptions.wic
        )
        expect(takeWebScreenshotSpy.mock.calls[0]).toMatchSnapshot()
        expect(afterScreenshotSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should call takeWebScreenshot with BiDi disabled when not available', async () => {
        canUseBidiScreenshotSpy.mockReturnValueOnce(false)
        const result = await saveWebScreen(baseOptions)

        expect(result).toMatchSnapshot()
        expect(takeWebScreenshotSpy.mock.calls[0]).toMatchSnapshot()
        expect(afterScreenshotSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should call takeWebScreenshot with BiDi disabled when mobile device', async () => {
        canUseBidiScreenshotSpy.mockReturnValueOnce(true)
        const beforeScreenshotMock = createBeforeScreenshotMock({ isMobile: true })
        vi.mocked((await import('../helpers/beforeScreenshot.js')).default).mockResolvedValueOnce(beforeScreenshotMock)

        const result = await saveWebScreen(baseOptions)

        expect(result).toMatchSnapshot()
        expect(takeWebScreenshotSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should call takeWebScreenshot with BiDi disabled when legacy method enabled', async () => {
        canUseBidiScreenshotSpy.mockReturnValueOnce(true)
        const options = createTestOptions(baseOptions, {
            saveScreenOptions: {
                ...baseOptions.saveScreenOptions,
                method: createMethodOptions({
                    enableLegacyScreenshotMethod: true
                })
            }
        })
        const result = await saveWebScreen(options)

        expect(result).toMatchSnapshot()
        expect(takeWebScreenshotSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should pass iOS configuration correctly', async () => {
        const beforeScreenshotMock = createBeforeScreenshotMock({
            isIOS: true,
            deviceName: 'iPhone 14 Pro',
            dimensions: {
                window: {
                    isLandscape: true
                }
            }
        })
        vi.mocked((await import('../helpers/beforeScreenshot.js')).default).mockResolvedValueOnce(beforeScreenshotMock)

        const options = createTestOptions(baseOptions, {
            saveScreenOptions: {
                ...baseOptions.saveScreenOptions,
                wic: {
                    ...baseOptions.saveScreenOptions.wic,
                    addIOSBezelCorners: true
                }
            }
        })
        const result = await saveWebScreen(options)

        expect(result).toMatchSnapshot()
        expect(takeWebScreenshotSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should handle NaN dimension values correctly', async () => {
        const nanDimensions = createBeforeScreenshotMock({
            dimensions: {
                window: {
                    devicePixelRatio: NaN,
                    innerHeight: NaN,
                    innerWidth: NaN,
                    isEmulated: false,
                    isLandscape: false,
                    outerHeight: NaN,
                    outerWidth: NaN,
                    screenHeight: NaN,
                    screenWidth: NaN,
                },
            },
            devicePixelRatio: NaN,
            initialDevicePixelRatio: NaN
        })
        vi.mocked((await import('../helpers/beforeScreenshot.js')).default).mockResolvedValueOnce(nanDimensions)
        const result = await saveWebScreen(baseOptions)

        expect(result).toMatchSnapshot()
        expect(takeWebScreenshotSpy.mock.calls[0]).toMatchSnapshot()
        expect(afterScreenshotSpy.mock.calls[0]).toMatchSnapshot()
    })
})
