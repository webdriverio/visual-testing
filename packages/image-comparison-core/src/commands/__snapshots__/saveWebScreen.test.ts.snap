// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`saveWebScreen > should call takeWebScreenshot with BiDi disabled when legacy method enabled 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-screen.png",
}
`;

exports[`saveWebScreen > should call takeWebScreenshot with BiDi disabled when legacy method enabled 2`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addIOSBezelCorners": false,
    "deviceName": "desktop",
    "devicePixelRatio": 2,
    "enableLegacyScreenshotMethod": true,
    "initialDevicePixelRatio": 2,
    "innerHeight": 900,
    "innerWidth": 1200,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isEmulated": false,
    "isIOS": false,
    "isLandscape": false,
    "isMobile": false,
  },
  false,
]
`;

exports[`saveWebScreen > should call takeWebScreenshot with BiDi disabled when mobile device 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-screen.png",
}
`;

exports[`saveWebScreen > should call takeWebScreenshot with BiDi disabled when mobile device 2`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addIOSBezelCorners": false,
    "deviceName": "desktop",
    "devicePixelRatio": 2,
    "enableLegacyScreenshotMethod": false,
    "initialDevicePixelRatio": 2,
    "innerHeight": 900,
    "innerWidth": 1200,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isEmulated": false,
    "isIOS": false,
    "isLandscape": false,
    "isMobile": true,
  },
  false,
]
`;

exports[`saveWebScreen > should call takeWebScreenshot with BiDi disabled when not available 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-screen.png",
}
`;

exports[`saveWebScreen > should call takeWebScreenshot with BiDi disabled when not available 2`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addIOSBezelCorners": false,
    "deviceName": "desktop",
    "devicePixelRatio": 2,
    "enableLegacyScreenshotMethod": false,
    "initialDevicePixelRatio": 2,
    "innerHeight": 900,
    "innerWidth": 1200,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isEmulated": false,
    "isIOS": false,
    "isLandscape": false,
    "isMobile": false,
  },
  false,
]
`;

exports[`saveWebScreen > should call takeWebScreenshot with BiDi disabled when not available 3`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "actualFolder": "/test/actual",
    "base64Image": "web-screenshot-data",
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "fileName": {
      "browserName": "chrome",
      "browserVersion": "120.0.0",
      "deviceName": "desktop",
      "devicePixelRatio": 2,
      "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
      "isMobile": false,
      "isTestInBrowser": true,
      "logName": "chrome",
      "name": "chrome",
      "outerHeight": 1000,
      "outerWidth": 1200,
      "platformName": "desktop",
      "platformVersion": "120.0.0",
      "screenHeight": 1080,
      "screenWidth": 1920,
      "tag": "test-screen",
    },
    "filePath": {
      "browserName": "chrome",
      "deviceName": "desktop",
      "isMobile": false,
      "savePerInstance": true,
    },
    "hideElements": [],
    "hideScrollBars": true,
    "isLandscape": false,
    "isNativeContext": false,
    "platformName": "Windows 10",
    "removeElements": [],
  },
]
`;

exports[`saveWebScreen > should call takeWebScreenshot with correct options when BiDi is available 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-screen.png",
}
`;

exports[`saveWebScreen > should call takeWebScreenshot with correct options when BiDi is available 2`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addIOSBezelCorners": false,
    "deviceName": "desktop",
    "devicePixelRatio": 2,
    "enableLegacyScreenshotMethod": false,
    "initialDevicePixelRatio": 2,
    "innerHeight": 900,
    "innerWidth": 1200,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isEmulated": false,
    "isIOS": false,
    "isLandscape": false,
    "isMobile": false,
  },
  true,
]
`;

exports[`saveWebScreen > should call takeWebScreenshot with correct options when BiDi is available 3`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "actualFolder": "/test/actual",
    "base64Image": "web-screenshot-data",
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "fileName": {
      "browserName": "chrome",
      "browserVersion": "120.0.0",
      "deviceName": "desktop",
      "devicePixelRatio": 2,
      "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
      "isMobile": false,
      "isTestInBrowser": true,
      "logName": "chrome",
      "name": "chrome",
      "outerHeight": 1000,
      "outerWidth": 1200,
      "platformName": "desktop",
      "platformVersion": "120.0.0",
      "screenHeight": 1080,
      "screenWidth": 1920,
      "tag": "test-screen",
    },
    "filePath": {
      "browserName": "chrome",
      "deviceName": "desktop",
      "isMobile": false,
      "savePerInstance": true,
    },
    "hideElements": [],
    "hideScrollBars": true,
    "isLandscape": false,
    "isNativeContext": false,
    "platformName": "Windows 10",
    "removeElements": [],
  },
]
`;

exports[`saveWebScreen > should handle NaN dimension values correctly 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-screen.png",
}
`;

exports[`saveWebScreen > should handle NaN dimension values correctly 2`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addIOSBezelCorners": false,
    "deviceName": "desktop",
    "devicePixelRatio": 1,
    "enableLegacyScreenshotMethod": false,
    "initialDevicePixelRatio": NaN,
    "innerHeight": NaN,
    "innerWidth": NaN,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isEmulated": false,
    "isIOS": false,
    "isLandscape": false,
    "isMobile": false,
  },
  false,
]
`;

exports[`saveWebScreen > should handle NaN dimension values correctly 3`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "actualFolder": "/test/actual",
    "base64Image": "web-screenshot-data",
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "fileName": {
      "browserName": "chrome",
      "browserVersion": "120.0.0",
      "deviceName": "desktop",
      "devicePixelRatio": NaN,
      "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
      "isMobile": false,
      "isTestInBrowser": true,
      "logName": "chrome",
      "name": "chrome",
      "outerHeight": NaN,
      "outerWidth": NaN,
      "platformName": "desktop",
      "platformVersion": "120.0.0",
      "screenHeight": NaN,
      "screenWidth": NaN,
      "tag": "test-screen",
    },
    "filePath": {
      "browserName": "chrome",
      "deviceName": "desktop",
      "isMobile": false,
      "savePerInstance": true,
    },
    "hideElements": [],
    "hideScrollBars": true,
    "isLandscape": false,
    "isNativeContext": false,
    "platformName": "Windows 10",
    "removeElements": [],
  },
]
`;

exports[`saveWebScreen > should pass iOS configuration correctly 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-screen.png",
}
`;

exports[`saveWebScreen > should pass iOS configuration correctly 2`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addIOSBezelCorners": true,
    "deviceName": "iPhone 14 Pro",
    "devicePixelRatio": 1,
    "enableLegacyScreenshotMethod": false,
    "initialDevicePixelRatio": 2,
    "innerHeight": undefined,
    "innerWidth": undefined,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isEmulated": undefined,
    "isIOS": true,
    "isLandscape": true,
    "isMobile": false,
  },
  false,
]
`;
